### PR TITLE
[SYCL][L0] Add SYCL_PI_LEVEL_ZERO_USM_RESIDENT to control if/where to make USM allocation resident

### PR DESCRIPTION
**Pull request by smaslov-intel**

### CODE REVIEW

**Comment by abagusetty** [comment]
@smaslov-intel Was just wondering, if one uses `SYCL_ENABLE_DEFAULT_CONTEXTS=1` that enables a default context for all devices on the platform. Will this variable interfere with `SYCL_PI_LEVEL_ZERO_USM_RESIDENT=1` in any way because the default-context is same for other devices. 

Also this `SYCL_PI_LEVEL_ZERO_USM_RESIDENT` var applies to both per-tile (explicit-scaling) and per-device (implicit-scaling) ? 
Should one expect any performance improvement with this residency

**Comment by smaslov-intel** [comment]
@abagusetty:

> @smaslov-intel Was just wondering, if one uses `SYCL_ENABLE_DEFAULT_CONTEXTS=1` that enables a default context for all devices on the platform. Will this variable interfere with `SYCL_PI_LEVEL_ZERO_USM_RESIDENT=1` in any way because the default-context is same for other devices.

The "default" context will be the same, and will have contain all root-devices of the platform.
Now, with `SYCL_PI_LEVEL_ZERO_USM_RESIDENT=2` setting would make allocations resident on all those root-devices.
So accesses to those allocation (direct or indirect) will become faster since allocations are already resident.

> Also this `SYCL_PI_LEVEL_ZERO_USM_RESIDENT` var applies to both per-tile (explicit-scaling) and per-device (implicit-scaling) ? 

The `SYCL_PI_LEVEL_ZERO_USM_RESIDENT=2` applies to all devices in the context that you create (or default)
No other diff between explicit vs. implicit.

> Should one expect any performance improvement with this residency

Somewhat slower allocations (that will now involve making them resident), but faster accesses later on.
Note, the default is `SYCL_PI_LEVEL_ZERO_USM_RESIDENT=0` which is no change from the current behavior.

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 1]
```diff
@@ -6864,12 +6864,73 @@ pi_result piextGetDeviceFunctionPointer(pi_device Device, pi_program Program,
...
+  } else {
+    Devices.push_back(Device);
+    if (USMAllocationForceResidency ==
+        USMAllocationForceResidencyType::P2PDevices) {
+      ze_bool_t P2P;
+      for (const auto &D : Context->Devices) {
+        if (D == Device)
+          continue;
+        // TODO: Cache P2P devices for a context
+        ZE_CALL(zeDeviceCanAccessPeer, (D->ZeDevice, Device->ZeDevice, &P2P));
```
Comment (jandres742) [inline sycl/plugins/level_zero/pi_level_zero.cpp:6922]: this is already cached in L0 as well, so cost of calling canAccessPeer is optimized in next call.

**Review by jandres742 — COMMENTED** [review]
looking good.

**Comment by jandres742** [comment]
> Also this `SYCL_PI_LEVEL_ZERO_USM_RESIDENT` var applies to both per-tile (explicit-scaling) and per-device (implicit-scaling) ?
> Should one expect any performance improvement with this residency

Allocations are made resident in the device that owns it. If using implicit scaling and allocation was created against the root device, then allocation is made resident in the root device, and that means resident for all tiles. If allocation was created instead against a sub-device, then the allocation needs to be made resident on the sub-device and siblings.

@smaslov: we are missing that: making it resident in the sub-device siblings.

**Comment by jandres742** [comment]
> > Also this `SYCL_PI_LEVEL_ZERO_USM_RESIDENT` var applies to both per-tile (explicit-scaling) and per-device (implicit-scaling) ?
> > Should one expect any performance improvement with this residency
> 
> Allocations are made resident in the device that owns it. If using implicit scaling and allocation was created against the root device, then allocation is made resident in the root device, and that means resident for all tiles. If allocation was created instead against a sub-device, then the allocation needs to be made resident on the sub-device and siblings.
> 
> @smaslov: we are missing that: making it resident in the sub-device siblings.

talked with @smaslov-intel , this is covered with =2. thanks!

**Review by jandres742 — COMMENTED** [review]
+1

**Comment by bader** [comment]
@smaslov-intel, please, merge head of `sycl` branch to your private branch if you want to have CI checks green. It's not necessary for this PR as all L0 tests pass, but CUDA checks won't return.

**Comment by smaslov-intel** [comment]
> @smaslov-intel, please, merge head of `sycl` branch to your private branch if you want to have CI checks green. It's not necessary for this PR as all L0 tests pass, but CUDA checks won't return.

The CUDA plugin is not affected by the changes in the PR, so why spend resources re-testing it.
@intel/llvm-gatekeepers : please help to merge this one.

**Review by againull — APPROVED** [review]

**Pull request merged**